### ISSUE TITLE
Is "operate" too specific?

### ISSUE THREAD

**Issue by kemitchell** [licensezero/parity-public-license#67]
Pulling [@makkus' comment](https://github.com/licensezero/parity-public-license/issues/57#issuecomment-544730196) out into its own issue.

To try and paraphrase: The verb "operate" implies a bit more than just "run".

If program A invokes program B, it's pretty easy to say that an end-user uses A to operate B. But it may not seem right to say that the end-user also uses B to operate A.

@makkus, I think a concrete example would help here. Did you have one in mind?

**Comment by kemitchell** [comment]
@makkus, still very interested in your example, if you have one. But two thoughts occurred to me.

First, I think the answer to one or more specific examples might be in "develop". Whether a users uses A to "operate" B or not, if the user uses A to "develop" B, they still have to contribute B.

Second, the fix here, if indeed a fix is needed, might be to continue using "operate", but make sure it goes both ways:

> ## Copyleft
>
> Contribute software you develop, operate, or analyze with this software, including changes or additions to this software, <ins>as well as software you use to develop, operate, or analyze this software</ins>. When in doubt, contribute.

In other words: Make sure each verb in "contribute" goes both ways.

But I'm not sure that's "free for open source"!

Let's focus on specific examples if we can. This is getting pretty abstract.

**Comment by jeremy-w** [comment]
Including the reverse language about having to contribute the things I use to develop the software  seems like it might rule out turning on Github’s dependabot security analysis of repos? It helps me develop the Parity-licensed software, but it’s magic server-side code I can’t do anything with myself.

Or what about closed-source editors like Sublime Text or Textmate?

**Comment by jeremy-w** [comment]
FWIW I read “operate” as something like “deploy, orchestrate, and monitor”.

If I had Parity-licensed daemontools, all the services and service files it’s sparking would be operated by it. If I had Parity-licensed Xen VMM, everything above it in the stack would need to have been contributed - no Windows or macOS VMs, and careful with those drivers, since the VMM is managing all those VMs.

If someone else is operating my Parity-licensed MongoDB, I wouldn’t expect their operations codes for operating the DB to be required to be contributed: they aren’t using the DB to operate things, but rather are operating the DB.

I’m not even clear on what the desired outcome would be for a Parity-licensed MongoDB in that case; maybe disallowing keeping secret that ops code is what’s expected, but it’s not what I’d expect with the current verbiage.

**Comment by kemitchell** [comment]
@jeremy-w I read "operate" much as you do, with the possible addition of "run", so:

> any combination of run, deploy, orchestrate, and monitor

But the point of "operate" is exactly how expansive it is, and drafting off its extensive use in trade for that very breadth.

I also trace the daemontools and Xen "test cases" as you do. It is absolutely the case that Parity will go too far for some kinds of software. The running example thus far has been a text editor: developing a program with a Parity-licensed Emacs would require contributing the program. But Parity leaves the choice of whether that's advisable up to the developer or company picking the license.

From my own point of view, the primary point of choosing Parity for a database like MongoDB would be to require all applications built with the database be open source, plus changes, extensions, forks, and so on.

**Comment by makkus** [comment]
So, concrete examples would be user-interfaces, or wrapper tools. Say 'flake8' (a Python syntax checker/linter) would be Parity licensed. All code it would analyze would have to be open. But if I wrote a proprietary user interface for it? flake8 itself wouldn't 'operate' on any non-open code, but it would be 'operated by' one. As I mentioned in my other commend, something like 'operate in combination with' would probably work. In this example we also 'develop' and 'analyze', but again, we do it 'downstream'. 'Upstream' we just execute our Parity-licensed code. Does that make sense?

If we want to take 'develop' and 'analyze' out of the equation, we could imagine 'Ansible' to be Paritly licensed as an example. If I write a user interface for it, or another tool that uses Ansible downstream, it would be the same: Ansible operates on other software, installs systems, etc. Our GUI/wrapper would operate Ansible, but could we say that Ansible operates our wrapper software?

**Comment by kemitchell** [comment]
@makkus running with your flake8 example:

If I build a user interface for flake8 or Ansible, is it fair to say that I've developed that interface with flake8 or Ansible?

**Comment by makkus** [comment]
Yes, probably. I was more thinking of the situation where you develop the user interface, then sell it to me (using a non-free license), and I use the whole shebang on free software. So, I'd use your software, which would operate the Parity-licensed code, which in turn would operate on free software.

**Comment by kemitchell** [comment]
In the case of a linter like flake8, end users develop and analyze software with the linter. Whether the whole linter is available under Parity, or just a library within it, Parity requires contributing linted software.

In the case of an orchestrator like Ansible, end users operate software with the orchestrator. Whether the whole orchestrator is available under Parity, or just a library within it, Parity requires contributing orchestrated software.

Recall that Parity explicitly does _not_ allow changing the license terms for Parity-licensed code:

> ## Contribute
> ... Note that this license does _not_ allow you to change the license terms for this software.

**Comment by makkus** [comment]
Sorry, this is probably one of those situations where I can't see the forest for the trees, and I'm missing something really obvious.

So, yes, I think I understand your argument, but I'm not sure it applies. There is still one 'non-free' software in the pipeline (at the beginning of it): the wrapper/GUI code. So, just to recap:

- I write Ansible, and license it using Parity
- you write a proprietary GUI for Ansible

So far, this means:
- you have to get a private license from me, because you develop software with my code

Now:
- you sell your GUI to a 3rd party
- 3rd party does not get a private license from me for Ansible
- 3rd party uses your GUI to develop/use 100% free software (apart from GUI)

The thing I'm not certain about:
- would this be ok, or would 3rd party have to get a private license for Ansible?
- this is where my reading of 'operate' comes in: does Ansible actually 'operate' any non-free software in this case, since it is in the middle of the pipeline, and only 'touches' open software downstream, although it is 'being touched' by non-free software upstream.

If this situation does not trigger Parity: do we want it to, or do we not?

**Comment by kemitchell** [comment]
@makkus in the hypothetical you sketched, with no additional facts, using a mix of Parity-licensed Ansible and other-licensed GUI to deploy a closed application would trigger the Parity copyleft rule for Parity-licensed Ansible, in my view. The license terms for Parity-licensed Ansible don't change because they've been wrapped with a GUI. The developer of the GUI must follow Notices and give everyone who receives a copy notice of the Parity terms that apply to it.

Note, however, that the private license sold to the GUI developer _could_ allow them to sublicense without copyleft. License Zero's standard-form private license explicitly does _not_ allow this:

> ## Sublicensing
> 
>   If you combine my contributions to the project with other software in a larger application, you may sublicense my contributions to the project as part of your larger application, and allow further sublicensing in turn, under these rules:
> 
> ...
> 
>   2. You may not sublicense anyone to break any rule of the public license for my contributions to the project for any changes of their own or any software besides your larger application.
> 
> ...

**Comment by makkus** [comment]
Yes, but the 3rd party does not deploy a closed application, but a 100% free one. The GUI itself is not 'deployed', it just controls Ansible which does the deploying of free code.

**Comment by makkus** [comment]
Cool, thanks!

**Comment by kemitchell** [comment]
Thank _you_, @makkus. This was very illuminating for me!

**Comment by makkus** [comment]
Right back at you, for me too :-)
